File: headless/src/main/org/nlogo/prim/_inlinkneighbor.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim;

import org.nlogo.agent.AgentSet;
import org.nlogo.agent.Turtle;
import org.nlogo.api.LogoException;
import org.nlogo.api.Syntax;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.Reporter;

public final strictfp class _inlinkneighbor
    extends Reporter {
  private final String breedName;

  public _inlinkneighbor() {
    breedName = null;
  }

  public _inlinkneighbor(String breedName) {
    this.breedName = breedName;
  }

  @Override
  public Syntax syntax() {
    return Syntax.reporterSyntax
        (new int[]{Syntax.AgentType()},
            Syntax.BooleanType(), "-T--");
  }

  @Override
  public String toString() {
    return super.toString() + ":" + breedName;
  }

  @Override
  public Object report(final Context context) throws LogoException {
    Turtle target = argEvalTurtle(context, 0);
    AgentSet breed =
        breedName == null
            ? world.links()
            : world.getLinkBreed(breedName);
    mustNotBeUndirected(breed, context);
    return null == world.linkManager.findLinkFrom(target, (Turtle) context.agent,
        breed, true)
        ? Boolean.FALSE
        : Boolean.TRUE;
  }
}

File: netlogo-gui/src/main/prim/etc/_inlinkneighbor.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.etc

import org.nlogo.agent.{AgentSet, Turtle}
import org.nlogo.nvm.{Context, Reporter}

class _inlinkneighbor(breedName: String) extends Reporter {
  def this() = this(null)

  override def toString: String = s"${super.toString}:$breedName"

  override def report(context: Context) = Boolean.box(report_1(context, argEvalTurtle(context, 0)))

  def report_1(context: Context, target: Turtle): Boolean = {
    val parent: Turtle = context.agent.asInstanceOf[Turtle]
    val breed: AgentSet = if (breedName == null) world.links else world.getLinkBreed(breedName)
    world.linkManager.isLinkedTo(target, parent, breed)
  }
}
